### ISSUE TITLE
COSMIC (v90) support

### ISSUE THREAD

**Issue by jxshi** [bcbio/bcbio-nextgen#2932]
Hi,

```COSMIC``` database has updated to v90 today. It is stated that ```Release v90 (5 September 2019) includes important changes```. Maybe the ```prepare_cosmic.py``` script needs an update too because I got an error message when trying to update ```COSMIC``` database using the following command:
```
bcbio_python prepare_cosmic.py 90 ~/local/share/bcbio
```
Here is the error message:
```
2019-09-05 22:48:15,245 [I] Beginning COSMIC v90 prep for GRCh37.
2019-09-05 22:48:15,246 [I] v90/GRCh37 files exist, please use the --clean flag to overwrite the existing files if you want to reinstall.
2019-09-05 22:48:15,246 [I] Downloading COSMIC VCF files.
2019-09-05 22:48:15,246 [I] Downloading https://cancer.sanger.ac.uk/cosmic/file_download/GRCh37/cosmic/v90/VCF/CosmicCodingMuts.vcf.gz
Traceback (most recent call last):
  File "prepare_cosmic.py", line 220, in <module>
    main(args.cosmic_version, args.bcbio_directory, args.overwrite, args.clean)
  File "prepare_cosmic.py", line 56, in main
    cosmic_vcf_files = get_cosmic_vcf_files(genome_build, cosmic_version, clean)
  File "prepare_cosmic.py", line 200, in get_cosmic_vcf_files
    download_url = r.json()["url"]
KeyError: **'url'**
```

Thank you for your time!

Best,
Jianxiang

**Comment by roryk** [comment]
Hi Jianxiang,

This new release of COSMIC changed a bunch of stuff, so I'd hold off updating for a while.

**Comment by roryk** [comment]
Did you set your COSMIC_USER and COSMIC_PASS environment variables before running the script?

**Comment by jxshi** [comment]
> Did you set your COSMIC_USER and COSMIC_PASS environment variables before running the script?

Yes, I set these two environment variables before running the script.

Thank you!

**Comment by roryk** [comment]
Looks like we might have to rework the script-- a bunch of other stuff changed in COSMIC 90 too, so I'm guessing that it will also break some things downstream. So I'd hold off for now until we can test it out further.

**Comment by jxshi** [comment]
> Looks like we might have to rework the script-- a bunch of other stuff changed in COSMIC 90 too, so I'm guessing that it will also break some things downstream. So I'd hold off for now until we can test it out further.

Sounds great! Thanks and looking forward to the new update.

**Comment by roryk** [comment]
Thanks-- this works fine to me, I think maybe COSMIC hadn't updated their files yet. If you try again, does it work? If not there might be a firewall issue on your end or something similar to that. Leaving this issue open for now while I run a test bcbio run with v90 and seeing if it all looks good.

**Comment by roryk** [comment]
Tested on https://github.com/bcbio/bcbio_validation_workflows/tree/master/somatic-giab-mix and we mark the same variants as v89, so I think this works fine. We're just marking up the VCF files with the cosmic_id, so we should be insulated somewhat from the changes in format.

**Comment by chapmanb** [comment]
Konstantinos;
You need to download this script:
```
wget https://raw.githubusercontent.com/chapmanb/cloudbiolinux/master/utils/prepare_cosmic.py
```
See the full documentation on adding COSMIC here: https://bcbio-nextgen.readthedocs.io/en/latest/contents/installation.html#customizing-data-installation

Hope this helps.

**Comment by kokyriakidis** [comment]
My bad! Did not notice that I had to download it! Thanks and sorry!

**Comment by wangpenhok** [comment]
hello , when I try to configure my hg38 v97 COSMIC files using prepare_cosmic.py. Both vcf files were successfully downloaded, but it seems some changes have occurred to the vcf files. Specific details are as follows: 

```
2023-02-13 14:01:55,149 [I] Beginning COSMIC v97 prep for GRCh37.
2023-02-13 14:01:55,149 [I] Beginning COSMIC v97 prep for GRCh38.
2023-02-13 14:01:55,149 [I] Downloading COSMIC VCF files.
2023-02-13 14:01:55,149 [I] Downloading https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v97/VCF/CosmicCodingMuts.vcf.gz
2023-02-13 14:04:48,452 [I] Downloading https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v97/VCF/CosmicNonCodingVariants.vcf.gz
2023-02-13 14:05:57,476 [I] Sorting v97/GRCh38/CosmicCodingMuts.vcf.gz to match the order of /home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.fa.
> gsort version 0.1.4
[W::vcf_parse] Contig 'chr1' is not defined in the header. (Quick workaround: index the file with tabix.)
[E::vcf_format] Invalid BCF, CONTIG id=0 not present in the header
[flush_buffer] Error: cannot write to -
Failed to read from standard input: unknown file type
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x546d43]

goroutine 1 [running]:
bufio.NewReaderSize(...)
	/home/brentp/go/src/bufio/bufio.go:49
bufio.NewReader(...)
	/home/brentp/go/src/bufio/bufio.go:62
github.com/brentp/gsort.Sort(0x7ad380, 0x0, 0x7ad3a0, 0xc000670040, 0xc000670000, 0xaf0, 0x0, 0x0, 0x0)
	/home/brentp/go/go/src/github.com/brentp/gsort/gsort.go:116 +0x63
main.main()
	/home/brentp/go/go/src/github.com/brentp/gsort/cmd/gsort/gsort.go:373 +0x513
Failed to read from standard input: unknown file type
2023-02-13 14:05:57,619 [I] bgzipping and indexing v97/GRCh38/CosmicCodingMuts-prep.vcf.gz.
2023-02-13 14:05:57,678 [I] Sorting v97/GRCh38/CosmicNonCodingVariants.vcf.gz to match the order of /home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.fa.
> gsort version 0.1.4
[W::vcf_parse] Contig 'chr1' is not defined in the header. (Quick workaround: index the file with tabix.)
[E::vcf_format] Invalid BCF, CONTIG id=0 not present in the header
[flush_buffer] Error: cannot write to -
Failed to read from standard input: unknown file type
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x546d43]

goroutine 1 [running]:
bufio.NewReaderSize(...)
	/home/brentp/go/src/bufio/bufio.go:49
bufio.NewReader(...)
	/home/brentp/go/src/bufio/bufio.go:62
github.com/brentp/gsort.Sort(0x7ad380, 0x0, 0x7ad3a0, 0xc0005f5d40, 0xc0005f5d00, 0xaf0, 0x0, 0x0, 0x0)
	/home/brentp/go/go/src/github.com/brentp/gsort/gsort.go:116 +0x63
main.main()
	/home/brentp/go/go/src/github.com/brentp/gsort/cmd/gsort/gsort.go:373 +0x513
Failed to read from standard input: unknown file type
2023-02-13 14:05:57,761 [I] bgzipping and indexing v97/GRCh38/CosmicNonCodingVariants-prep.vcf.gz.
2023-02-13 14:05:57,798 [I] Combining COSMIC files to v97/bcbio_ready/hg38/cosmic.vcf.gz.
INFO	2023-02-13 14:05:59	MergeVcfs	

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** 
https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    MergeVcfs -O v97/bcbio_ready/hg38/cosmic.vcf.gz -D /home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.dict -I v97/GRCh38/CosmicCodingMuts-prep.vcf.gz -I v97/GRCh38/CosmicNonCodingVariants-prep.vcf.gz -USE_JDK_DEFLATER true -USE_JDK_INFLATER true -CREATE_INDEX false
**********


[Mon Feb 13 14:06:00 CST 2023] MergeVcfs INPUT=[v97/GRCh38/CosmicCodingMuts-prep.vcf.gz, v97/GRCh38/CosmicNonCodingVariants-prep.vcf.gz] OUTPUT=v97/bcbio_ready/hg38/cosmic.vcf.gz SEQUENCE_DICTIONARY=/home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.dict CREATE_INDEX=false USE_JDK_DEFLATER=true USE_JDK_INFLATER=true    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json
[Mon Feb 13 14:06:00 CST 2023] Executing as bcbio@dell-PowerEdge-R7525 on Linux 5.15.0-46-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_332-b09; Deflater: Jdk; Inflater: Jdk; Provider GCS is not available; Picard version: 2.27.5
[Mon Feb 13 14:06:01 CST 2023] picard.vcf.MergeVcfs done. Elapsed time: 0.00 minutes.
Runtime.totalMemory()=514850816
To get help, see http://broadinstitute.github.io/picard/index.html#GettingHelp
Exception in thread "main" htsjdk.tribble.TribbleException$MalformedFeatureFile: Unable to parse header with error: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file, for input source: file:///home/data/bcbio/cosmic-prep/v97/GRCh38/CosmicCodingMuts-prep.vcf.gz
	at htsjdk.tribble.TabixFeatureReader.readHeader(TabixFeatureReader.java:97)
	at htsjdk.tribble.TabixFeatureReader.<init>(TabixFeatureReader.java:82)
	at htsjdk.tribble.AbstractFeatureReader.getFeatureReader(AbstractFeatureReader.java:117)
	at htsjdk.tribble.AbstractFeatureReader.getFeatureReader(AbstractFeatureReader.java:81)
	at htsjdk.variant.vcf.VCFFileReader.<init>(VCFFileReader.java:145)
	at picard.vcf.MergeVcfs.doWork(MergeVcfs.java:182)
	at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:309)
	at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:103)
	at picard.cmdline.PicardCommandLine.main(PicardCommandLine.java:113)
Caused by: htsjdk.tribble.TribbleException$InvalidHeader: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file
	at htsjdk.variant.vcf.VCFCodec.readActualHeader(VCFCodec.java:119)
	at htsjdk.tribble.AsciiFeatureCodec.readHeader(AsciiFeatureCodec.java:79)
	at htsjdk.tribble.AsciiFeatureCodec.readHeader(AsciiFeatureCodec.java:37)
	at htsjdk.tribble.TabixFeatureReader.readHeader(TabixFeatureReader.java:95)
	... 8 more
Traceback (most recent call last):
  File "prepare_cosmic.py", line 246, in <module>
    main(args.cosmic_version, args.bcbio_directory, args.overwrite, args.clean)
  File "prepare_cosmic.py", line 62, in main
    ready_cosmic = combine_cosmic(sorted_inputs, bcbio_ref, out_file)
  File "prepare_cosmic.py", line 153, in combine_cosmic
    subprocess.check_call(cmd)
  File "/home/data/bcbio/anaconda/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['picard', 'MergeVcfs', 'O=v97/bcbio_ready/hg38/cosmic.vcf.gz', 'D=/home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.dict', 'I=v97/GRCh38/CosmicCodingMuts-prep.vcf.gz', 'I=v97/GRCh38/CosmicNonCodingVariants-prep.vcf.gz', 'USE_JDK_DEFLATER=true', 'USE_JDK_INFLATER=true', 'CREATE_INDEX=false']' returned non-zero exit status 1.
```

so i think updates of this script is needed.

**Comment by wangpenhok** [comment]
> hello , when I try to configure my hg38 v97 COSMIC files using prepare_cosmic.py. Both vcf files were successfully downloaded, but it seems some changes have occurred to the vcf files. Specific details are as follows:
> 
> ```
> 2023-02-13 14:01:55,149 [I] Beginning COSMIC v97 prep for GRCh37.
> 2023-02-13 14:01:55,149 [I] Beginning COSMIC v97 prep for GRCh38.
> 2023-02-13 14:01:55,149 [I] Downloading COSMIC VCF files.
> 2023-02-13 14:01:55,149 [I] Downloading https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v97/VCF/CosmicCodingMuts.vcf.gz
> 2023-02-13 14:04:48,452 [I] Downloading https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v97/VCF/CosmicNonCodingVariants.vcf.gz
> 2023-02-13 14:05:57,476 [I] Sorting v97/GRCh38/CosmicCodingMuts.vcf.gz to match the order of /home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.fa.
> > gsort version 0.1.4
> [W::vcf_parse] Contig 'chr1' is not defined in the header. (Quick workaround: index the file with tabix.)
> [E::vcf_format] Invalid BCF, CONTIG id=0 not present in the header
> [flush_buffer] Error: cannot write to -
> Failed to read from standard input: unknown file type
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x546d43]
> 
> goroutine 1 [running]:
> bufio.NewReaderSize(...)
> 	/home/brentp/go/src/bufio/bufio.go:49
> bufio.NewReader(...)
> 	/home/brentp/go/src/bufio/bufio.go:62
> github.com/brentp/gsort.Sort(0x7ad380, 0x0, 0x7ad3a0, 0xc000670040, 0xc000670000, 0xaf0, 0x0, 0x0, 0x0)
> 	/home/brentp/go/go/src/github.com/brentp/gsort/gsort.go:116 +0x63
> main.main()
> 	/home/brentp/go/go/src/github.com/brentp/gsort/cmd/gsort/gsort.go:373 +0x513
> Failed to read from standard input: unknown file type
> 2023-02-13 14:05:57,619 [I] bgzipping and indexing v97/GRCh38/CosmicCodingMuts-prep.vcf.gz.
> 2023-02-13 14:05:57,678 [I] Sorting v97/GRCh38/CosmicNonCodingVariants.vcf.gz to match the order of /home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.fa.
> > gsort version 0.1.4
> [W::vcf_parse] Contig 'chr1' is not defined in the header. (Quick workaround: index the file with tabix.)
> [E::vcf_format] Invalid BCF, CONTIG id=0 not present in the header
> [flush_buffer] Error: cannot write to -
> Failed to read from standard input: unknown file type
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x546d43]
> 
> goroutine 1 [running]:
> bufio.NewReaderSize(...)
> 	/home/brentp/go/src/bufio/bufio.go:49
> bufio.NewReader(...)
> 	/home/brentp/go/src/bufio/bufio.go:62
> github.com/brentp/gsort.Sort(0x7ad380, 0x0, 0x7ad3a0, 0xc0005f5d40, 0xc0005f5d00, 0xaf0, 0x0, 0x0, 0x0)
> 	/home/brentp/go/go/src/github.com/brentp/gsort/gsort.go:116 +0x63
> main.main()
> 	/home/brentp/go/go/src/github.com/brentp/gsort/cmd/gsort/gsort.go:373 +0x513
> Failed to read from standard input: unknown file type
> 2023-02-13 14:05:57,761 [I] bgzipping and indexing v97/GRCh38/CosmicNonCodingVariants-prep.vcf.gz.
> 2023-02-13 14:05:57,798 [I] Combining COSMIC files to v97/bcbio_ready/hg38/cosmic.vcf.gz.
> INFO	2023-02-13 14:05:59	MergeVcfs	
> 
> ********** NOTE: Picard's command line syntax is changing.
> **********
> ********** For more information, please see:
> ********** 
> https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
> **********
> ********** The command line looks like this in the new syntax:
> **********
> **********    MergeVcfs -O v97/bcbio_ready/hg38/cosmic.vcf.gz -D /home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.dict -I v97/GRCh38/CosmicCodingMuts-prep.vcf.gz -I v97/GRCh38/CosmicNonCodingVariants-prep.vcf.gz -USE_JDK_DEFLATER true -USE_JDK_INFLATER true -CREATE_INDEX false
> **********
> 
> 
> [Mon Feb 13 14:06:00 CST 2023] MergeVcfs INPUT=[v97/GRCh38/CosmicCodingMuts-prep.vcf.gz, v97/GRCh38/CosmicNonCodingVariants-prep.vcf.gz] OUTPUT=v97/bcbio_ready/hg38/cosmic.vcf.gz SEQUENCE_DICTIONARY=/home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.dict CREATE_INDEX=false USE_JDK_DEFLATER=true USE_JDK_INFLATER=true    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json
> [Mon Feb 13 14:06:00 CST 2023] Executing as bcbio@dell-PowerEdge-R7525 on Linux 5.15.0-46-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_332-b09; Deflater: Jdk; Inflater: Jdk; Provider GCS is not available; Picard version: 2.27.5
> [Mon Feb 13 14:06:01 CST 2023] picard.vcf.MergeVcfs done. Elapsed time: 0.00 minutes.
> Runtime.totalMemory()=514850816
> To get help, see http://broadinstitute.github.io/picard/index.html#GettingHelp
> Exception in thread "main" htsjdk.tribble.TribbleException$MalformedFeatureFile: Unable to parse header with error: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file, for input source: file:///home/data/bcbio/cosmic-prep/v97/GRCh38/CosmicCodingMuts-prep.vcf.gz
> 	at htsjdk.tribble.TabixFeatureReader.readHeader(TabixFeatureReader.java:97)
> 	at htsjdk.tribble.TabixFeatureReader.<init>(TabixFeatureReader.java:82)
> 	at htsjdk.tribble.AbstractFeatureReader.getFeatureReader(AbstractFeatureReader.java:117)
> 	at htsjdk.tribble.AbstractFeatureReader.getFeatureReader(AbstractFeatureReader.java:81)
> 	at htsjdk.variant.vcf.VCFFileReader.<init>(VCFFileReader.java:145)
> 	at picard.vcf.MergeVcfs.doWork(MergeVcfs.java:182)
> 	at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:309)
> 	at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:103)
> 	at picard.cmdline.PicardCommandLine.main(PicardCommandLine.java:113)
> Caused by: htsjdk.tribble.TribbleException$InvalidHeader: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file
> 	at htsjdk.variant.vcf.VCFCodec.readActualHeader(VCFCodec.java:119)
> 	at htsjdk.tribble.AsciiFeatureCodec.readHeader(AsciiFeatureCodec.java:79)
> 	at htsjdk.tribble.AsciiFeatureCodec.readHeader(AsciiFeatureCodec.java:37)
> 	at htsjdk.tribble.TabixFeatureReader.readHeader(TabixFeatureReader.java:95)
> 	... 8 more
> Traceback (most recent call last):
>   File "prepare_cosmic.py", line 246, in <module>
>     main(args.cosmic_version, args.bcbio_directory, args.overwrite, args.clean)
>   File "prepare_cosmic.py", line 62, in main
>     ready_cosmic = combine_cosmic(sorted_inputs, bcbio_ref, out_file)
>   File "prepare_cosmic.py", line 153, in combine_cosmic
>     subprocess.check_call(cmd)
>   File "/home/data/bcbio/anaconda/lib/python3.7/subprocess.py", line 363, in check_call
>     raise CalledProcessError(retcode, cmd)
> subprocess.CalledProcessError: Command '['picard', 'MergeVcfs', 'O=v97/bcbio_ready/hg38/cosmic.vcf.gz', 'D=/home/data/bcbio/genomes/Hsapiens/hg38/seq/hg38.dict', 'I=v97/GRCh38/CosmicCodingMuts-prep.vcf.gz', 'I=v97/GRCh38/CosmicNonCodingVariants-prep.vcf.gz', 'USE_JDK_DEFLATER=true', 'USE_JDK_INFLATER=true', 'CREATE_INDEX=false']' returned non-zero exit status 1.
> ```
> 
> so i think updates of this script is needed.

i have also tried v90, but it seems the v90 vcf files are not available from COSMIC anymore
